### PR TITLE
feat: use bwa-aln-interactive and upgrade developer's docs

### PR DESCRIPTION
**Pull request by clintval**
Closes https://github.com/fulcrumgenomics/prymer/issues/5
Closes https://github.com/fulcrumgenomics/prymer/issues/30
Closes https://github.com/fulcrumgenomics/prymer/issues/27

This PR also:

- Addresses improper installation of poetry.
- Adds some polish to the main README and developer documentation.

### CODE REVIEW

**Comment by clintval** [comment]
Local and remote tests hang forever when you install `bwa-aln-interactive` from bioconda and run tests.

The built executable from bioconda is either broken or this library isn't compatible with it yet.

Screenshot 2024-09-19 at 7 33 29 PM

**Comment by clintval** [comment]
I've narrowed the issue down to this "hack" which does not appear to work for me:

https://github.com/fulcrumgenomics/prymer/blob/e942368ca0e28d5f0533c6fc5a6b6b35b59cf764/prymer/offtarget/bwa.py#L287-L300

**Comment by clintval** [comment]
I figured it out. The GitHub release of `bwa-aln-interactive` is broken and we are shipping the non-interactive bwa. I'll next fix up the public repository and the bioconda package by incrementing the `build` number.

As a side quest, I was able to remove the `AlignmentFile` "hack"!

**Comment by msto** [comment]
> I wish prymer had better logging redirects for the underlying code it calls. Debugging this was painful without knowing where all the standard errors go (hint... it's /dev/null and the user can't change that).

Can you open a follow-up issue for this? I agree

**Comment by clintval** [comment]
> > I wish prymer had better logging redirects for the underlying code it calls. Debugging this was painful without knowing where all the standard errors go (hint... it's /dev/null and the user can't change that).
> 
> Can you open a follow-up issue for this? I agree

You have one open for Primer3:

- https://github.com/fulcrumgenomics/prymer/issues/15

I wrote one up for bwa:

- https://github.com/fulcrumgenomics/prymer/issues/41

File: prymer/offtarget/bwa.py [hunk 1]
```diff
@@ -297,10 +306,9 @@ def __init__(
 
     def __signal_bwa(self) -> None:
         """Signals BWA to process the queries"""
-        for _ in range(3):
-            self._subprocess.stdin.flush()
-            self._subprocess.stdin.write("\n\n")
-            self._subprocess.stdin.flush()
+        self._subprocess.stdin.flush()
+        self._subprocess.stdin.write("\n" * 16)
+        self._subprocess.stdin.flush()
```
Comment (clintval) [inline prymer/offtarget/bwa.py:311]: The Bioconda built executable requires more newlines before emitting an alignment.

Why the executable (regardless of compilation environment) needs more than 2 newlines is currently a mystery.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -19,87 +19,46 @@ jobs:
...
 
       - name: Set up Python ${{ matrix.PYTHON_VERSION }}
         uses: actions/setup-python@v5
         with:
           python-version: ${{ matrix.PYTHON_VERSION }}
 
+      - name: Install poetry
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install poetry==${{env.POETRY_VERSION}}
```
Comment (msto) [inline .github/workflows/tests.yml:31]: **suggestion** Should we use `pipx` to install `poetry` so it gets its own environment?

File: prymer/offtarget/bwa.py [hunk 3]
```diff
@@ -297,10 +306,9 @@ def __init__(
 
     def __signal_bwa(self) -> None:
         """Signals BWA to process the queries"""
-        for _ in range(3):
-            self._subprocess.stdin.flush()
-            self._subprocess.stdin.write("\n\n")
-            self._subprocess.stdin.flush()
+        self._subprocess.stdin.flush()
+        self._subprocess.stdin.write("\n" * 16)
```
Comment (msto) [inline prymer/offtarget/bwa.py:310]: **suggestion** Leave a comment/breadcrumb to document the reason for this and the choice of 16

**Review by msto — APPROVED** [review]

**Comment by clintval** [comment]
@msto's tested the developer docs and tests passed.

I think this is ready for review!

**Comment by clintval** [comment]
https://github.com/coderabbitai full review

File: prymer/offtarget/bwa.py [hunk 4]
```diff
@@ -296,11 +308,12 @@ def __init__(
...
     def __signal_bwa(self) -> None:
-        """Signals BWA to process the queries"""
-        for _ in range(3):
-            self._subprocess.stdin.flush()
-            self._subprocess.stdin.write("\n\n")
-            self._subprocess.stdin.flush()
+        """Signals BWA to process the queries."""
+        self._subprocess.stdin.flush()
+        # NB: the executable compiled on different platforms require a different number of newlines
+        # NB: it is not currently understood why this is, but 16 spaces seems to work for all tested
```
Comment (tfenne) [inline prymer/offtarget/bwa.py:314]:
```suggestion
        # NB: it is not currently understood why this is, but 16 newlines seems to work for all tested
```

**Review by tfenne — APPROVED** [review]

**Pull request merged**